### DISCUSSION TITLE
How to change automatic documentation generation schema name for Generic Pydantic datatypes?

### DISCUSSION THREAD

**Discussion by not-sekiun** [fastapi/fastapi#11129] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import TypeVar, Generic
from pydantic import BaseModel
from fastapi import FastAPI

T = TypeVar("T")


class GenericModel(BaseModel, Generic[T]):
    a: T


A = GenericModel[int]
B = GenericModel[str]
C = GenericModel[bool]

app = FastAPI()


@app.get("/", responses={400: {"model": A}, 401: {"model": B}, 402: {"model": C}})
def read_root() -> dict[str, str]:
    return {"Hello": "World"}
```

### Description

Running the above example code leads to the following generation of schema names at /openapi.json. Only the relevant fields are present, everything else is omitted.

```json
    "components": {
        "schemas": {
            "GenericModel_bool_": {...},
            "GenericModel_int_": {...},
            "GenericModel_str_": {...},
        }
    }
}
```

As you can see the generation of schema names for the GenericModel type takes the form of GenericModel__datatype__. I do not like this formatting, especially when it shows up in the swagger UI or redocs pages. I know that I can manually edit the openapi.json file and replace the names by retrieving its dictionary from app() but this can be tedious and ugly especially as the app grows lager, is there a better way to specify the automatic name generation format?

### Operating System

Windows

### FastAPI Version

0.104.1

### Pydantic Version

2.5.3

### Python Version

Python 3.12.0

### Additional Context

Schemas look like this on swagger UI

![Screenshot 2024-02-11 125221](https://github.com/tiangolo/fastapi/assets/32593795/4aa26390-31c9-45b4-859e-c22723b55d53)

**Comment by Vivanov98** [comment]
Another part of this same issue is that when you expand each generic model schema, the schema reference to the generic var doesn't seem to propogate properly (the generic var isn't named)

e.g. 
![image](https://github.com/tiangolo/fastapi/assets/66319645/450d7dab-d6e5-4d91-9d93-cb0d5ac7543a)

instead of 
![image](https://github.com/tiangolo/fastapi/assets/66319645/0f46d5b2-a61f-41b4-a2b2-ee778822eac9)


The work around I am looking to use for now is pinning the openAPI schema version
`app.openapi_version = "3.0.2"`

which produces the correct UI in Swagger (the former picture) though it is outdated

It is unclear to me if this is an issue with Swagger/OpenAPI or FastAPI needs to update its schema generation.
